MicroProfile OpenTracing 3.0

tonyreigns opened the pull request:
#17796

tonyreigns committed:
Initial drop of mpOT-3.0

tonyreigns committed:
Fixing DefaultExceptionMapper implemened classes

tonyreigns committed:
Fixing dependency issues

tonyreigns committed:
Debug messages

fmhwong commented on dev/io.openliberty.opentracing.2.0.internal/resources/META-INF/services/javax.ws.rs.ext.Providers:
```diff
@@ -0,0 +1,2 @@
+io.openliberty.opentracing.2.0.internal.OpentracingClientFilter
```
This should be `io.openliberty.opentracing.internal.OpentracingClientFilter`

fmhwong commented on dev/io.openliberty.opentracing.2.0.internal/resources/META-INF/services/javax.ws.rs.ext.Providers:
```diff
@@ -0,0 +1,2 @@
+io.openliberty.opentracing.2.0.internal.OpentracingClientFilter
+io.openliberty.opentracing.2.0.internal.OpentracingContainerFilter
```
This should be `io.openliberty.opentracing.internal.OpentracingContainerFilter`

tonyreigns replied:
Fixed but issue remains.

tonyreigns committed:
Fixing service loader classes

Azquelt committed:
Get opentracing 2.0 working again

OpenTracing 2.0 (for JAX-RS 2.1) still needs the JaxRsProviderRegister.
Split it out into a separate bundle which won't be used for OpenTracing
3.0.

Azquelt committed:
Initial OT3 fixes

Azquelt committed:
WIP OpenTracing 3.0 TCK

Currently it looks like there's a problem testing equality of responses
which is causing everything to fail.

Azquelt committed:
Switch to use resteasy when running the TCK

This seems to mostly work but still has a few failures.

skasund commented:
L2 message review not required.

fmhwong commented on dev/io.openliberty.microprofile.opentracing.3.0.internal_fat_tck/publish/tckRunner/pom.xml:
```diff
@@ -0,0 +1,71 @@
+    <name>MicroProfile Opentracing TCK Runner</name>
+
+    <properties>
+        <arquillian.version>1.3.0.Final</arquillian.version>
```
Use `1.7.0.Alpha9`

Azquelt committed:
restfulWS: Allow wrapping async client tasks

Allow the async client task to be wrapped. Initially, this is to allow
opentracing to propagate the currently active span to the request.

Azquelt committed:
opentracing-3.0: propagate span for async requests

Ensure the parent span is linked correctly by propagating the active
span to the new thread when making an asynchronous restfulWS request.

This is needed to accomodate a change in behaviour vs. JAX-RS 2.1 -
previously outgoing client filters ran on the thread that triggered the
request but now they run on a different thread.

jantley-ibm commented on dev/com.ibm.websphere.appserver.features/visibility/public/opentracing-3.0/resources/l10n/com.ibm.websphere.appserver.opentracing-3.0.properties:
```diff
@@ -0,0 +1,16 @@
+#NLS_MESSAGEFORMAT_NONE
+#
+
+description=This feature provides support for Opentracing function for enterprise Java.
```
Change
Opentracing function for enterprise Java.
to
the OpenTracing function for Enterprise Java.

tevans78 commented on dev/com.ibm.websphere.appserver.features/visibility/auto/io.openliberty.mpOpenTracing3.0-mpRestClient2.0.feature:
```diff
@@ -0,0 +1,10 @@
+  osgi.identity; filter:="(&(type=osgi.subsystem.feature)(osgi.identity=com.ibm.websphere.appserver.mpRestClient-2.0))"
+-bundles=io.openliberty.microprofile.opentracing.3.0.internal.rest.client.jakarta
+IBM-Install-Policy: when-satisfied
+kind=ga
```
Should be noship to start with

tevans78 commented on dev/com.ibm.websphere.appserver.features/visibility/auto/io.openliberty.mpOpenTracing3.0-mpRestClient2.0.feature:
```diff
@@ -0,0 +1,10 @@
+visibility=private
+IBM-Provision-Capability: \
+  osgi.identity; filter:="(&(type=osgi.subsystem.feature)(osgi.identity=com.ibm.websphere.appserver.mpOpenTracing-3.0))", \
+  osgi.identity; filter:="(&(type=osgi.subsystem.feature)(osgi.identity=com.ibm.websphere.appserver.mpRestClient-2.0))"
```
shouldn't this be `mpRestClient-3.0`?

tevans78 commented on dev/com.ibm.websphere.appserver.features/visibility/public/opentracing-3.0/com.ibm.websphere.appserver.opentracing-3.0.feature:
```diff
@@ -0,0 +1,29 @@
+         io.openliberty.opentracing.3.0.internal.restfulws
+-jars=io.openliberty.opentracing.2.0.spi; location:=dev/spi/ibm/
+-files= dev/spi/ibm/javadoc/io.openliberty.opentracing.2.0.spi_1.0-javadoc.zip
+kind=ga
```
noship

tonyreigns committed:
Version cleanup

tevans78 commented on dev/io.openliberty.microprofile.opentracing.3.0.internal_fat_tck/fat/src/io/openliberty/microprofile/opentracing/internal/tck/FATSuite.java:
```diff
@@ -0,0 +1,63 @@
+    
+
+    public static FeatureReplacementAction MP_REST_CLIENT(String version, String serverName) {
+        return use(new FeatureReplacementAction(), "mpRestClient", version, ALL_VERSIONS_OF_MP_REST_CLIENT)
```
You can't run this against anything other than mpRestClient-3.0

tonyreigns replied:
ALL_VERSIONS_OF_MP_REST_CLIENT is only set to 3.0 so shouldn't be an issue.

tevans78 commented on dev/com.ibm.websphere.appserver.features/visibility/public/mpOpenTracing-3.0/com.ibm.websphere.appserver.mpOpenTracing-3.0.feature:
```diff
@@ -0,0 +1,19 @@
+-include= ~${workspace}/cnf/resources/bnd/feature.props
+symbolicName=com.ibm.websphere.appserver.mpOpenTracing-3.0
```
Please name this `io.openliberty.mpOpenTracing-3.0`

tevans78 commented on dev/com.ibm.websphere.appserver.features/visibility/public/mpOpenTracing-3.0/com.ibm.websphere.appserver.mpOpenTracing-3.0.feature:
```diff
@@ -0,0 +1,19 @@
+    org.eclipse.microprofile.opentracing; type="stable"
+-features=com.ibm.websphere.appserver.opentracing-3.0, \
+  io.openliberty.mpConfig-3.0, \
+  com.ibm.websphere.appserver.org.eclipse.microprofile.opentracing-3.0, \
```
This should be `io.openliberty.org.eclipse.microprofile.opentracing-3.0`

tevans78 commented on dev/com.ibm.websphere.appserver.features/visibility/public/opentracing-3.0/com.ibm.websphere.appserver.opentracing-3.0.feature:
```diff
@@ -0,0 +1,29 @@
+-include= ~${workspace}/cnf/resources/bnd/feature.props
+symbolicName=com.ibm.websphere.appserver.opentracing-3.0
```
`io.openliberty.opentracing-3.0`

tevans78 commented on dev/com.ibm.websphere.appserver.features/visibility/auto/io.openliberty.mpOpenTracing3.0-mpRestClient3.0.feature:
```diff
@@ -0,0 +1,10 @@
+-include= ~${workspace}/cnf/resources/bnd/feature.props
+symbolicName=io.openliberty.mpOpenTracing3.0-mpRestClient2.0
```
```suggestion
symbolicName=io.openliberty.mpOpenTracing3.0-mpRestClient3.0
```

tevans78 commented on dev/com.ibm.websphere.appserver.features/visibility/auto/io.openliberty.mpOpenTracing3.0-mpRestClient3.0.feature:
```diff
@@ -0,0 +1,10 @@
+visibility=private
+IBM-Provision-Capability: \
+  osgi.identity; filter:="(&(type=osgi.subsystem.feature)(osgi.identity=com.ibm.websphere.appserver.mpOpenTracing-3.0))", \
+  osgi.identity; filter:="(&(type=osgi.subsystem.feature)(osgi.identity=com.ibm.websphere.appserver.mpRestClient-3.0))"
```
```suggestion
  osgi.identity; filter:="(&(type=osgi.subsystem.feature)(osgi.identity=io.openliberty.mpOpenTracing-3.0))", \
  osgi.identity; filter:="(&(type=osgi.subsystem.feature)(osgi.identity=io.openliberty.mpRestClient-3.0))"
```

tevans78 commented on dev/com.ibm.websphere.appserver.features/visibility/auto/io.openliberty.mpOpenTracing3.0-mpRestClient3.0.feature:
```diff
@@ -0,0 +1,10 @@
+  osgi.identity; filter:="(&(type=osgi.subsystem.feature)(osgi.identity=com.ibm.websphere.appserver.mpRestClient-3.0))"
+-bundles=io.openliberty.microprofile.opentracing.3.0.internal.rest.client.jakarta
+IBM-Install-Policy: when-satisfied
+kind=ga
```
noship

tevans78 commented on dev/com.ibm.websphere.appserver.features/visibility/private/com.ibm.websphere.appserver.org.eclipse.microprofile.opentracing-3.0.feature:
```diff
@@ -0,0 +1,9 @@
+-include= ~${workspace}/cnf/resources/bnd/feature.props
+symbolicName=com.ibm.websphere.appserver.org.eclipse.microprofile.opentracing-3.0
```
```suggestion
symbolicName=io.openliberty.org.eclipse.microprofile.opentracing-3.0
```

tevans78 commented on dev/com.ibm.websphere.appserver.features/visibility/public/mpOpenTracing-3.0/resources/l10n/com.ibm.websphere.appserver.mpOpenTracing-3.0.properties:
```diff
@@ -0,0 +1,18 @@
+#/*******************************************************************************
```
Don't forget to rename these files when you change the long feature name

tevans78 commented on dev/io.openliberty.microprofile.opentracing.3.0.internal_fat_tck/fat/src/io/openliberty/microprofile/opentracing/internal/tck/OpentracingTCKLauncherMicroProfile.java:
```diff
@@ -0,0 +1,56 @@
+    public void launchOpentracingTckMP() throws Exception {
+        // Use default tck-suite.xml
+        
+        MvnUtils.runTCKMvnCmd(server, "io.openliberty.opentracing.2.0.internal_fat", this.getClass() + ":launchOpentracingRestClientTck", "tck-and-rest-client-tck.xml", Collections.emptyMap(), Collections.emptySet());
```
```suggestion
        MvnUtils.runTCKMvnCmd(server, "io.openliberty.opentracing.3.0.internal_fat", this.getClass() + ":launchOpentracingRestClientTck", "tck-and-rest-client-tck.xml", Collections.emptyMap(), Collections.emptySet());
```

tevans78 commented on dev/io.openliberty.microprofile.opentracing.3.0.internal_fat_tck/publish/files/features/io.openliberty.opentracing.mock-3.0.mf:
```diff
@@ -0,0 +1,10 @@
+Subsystem-SymbolicName: io.openliberty.opentracing.mock-3.0; visibility:=public; singleton:=true
+Subsystem-Type: osgi.subsystem.feature
+Subsystem-Version: 1.0.0
+Subsystem-Content: com.ibm.websphere.appserver.opentracing-3.0; type="osgi.subsystem.feature",
```
```suggestion
Subsystem-Content: io.openliberty.opentracing-3.0; type="osgi.subsystem.feature",
```

tevans78 commented on dev/io.openliberty.microprofile.opentracing.3.0.internal_fat_tck/publish/servers/OpentracingTCKServerMicroProfile/server.xml:
```diff
@@ -0,0 +1,27 @@
+    Corporation - initial API and implementation -->
+<server>
+    <featureManager>
+        <feature>componenttest-1.0</feature>
```
```suggestion
        <feature>componenttest-2.0</feature>
```

tevans78 commented on dev/io.openliberty.microprofile.opentracing.3.0.internal_fat_tck/publish/tckRunner/pom.xml:
```diff
@@ -0,0 +1,71 @@
+    <dependencyManagement>
+        <dependencies>
+            <dependency>
+                <groupId>javax.enterprise</groupId>
```
should be jakarta cdi 3.0 equivalent

tevans78 commented on dev/io.openliberty.microprofile.opentracing.3.0.internal_fat_tck/publish/servers/OpentracingTCKServer/server.xml:
```diff
@@ -0,0 +1,25 @@
+    Corporation - initial API and implementation -->
+<server>
+    <featureManager>
+        <feature>componenttest-1.0</feature>
```
```suggestion
        <feature>componenttest-2.0</feature>
```

tevans78 commented on dev/io.openliberty.microprofile.opentracing.3.0.internal_fat_tck/publish/servers/OpentracingRestClientTCKServer/server.xml:
```diff
@@ -0,0 +1,26 @@
+    Corporation - initial API and implementation -->
+<server>
+    <featureManager>
+        <feature>componenttest-1.0</feature>
```
```suggestion
        <feature>componenttest-2.0</feature>
```

tevans78 commented on dev/io.openliberty.org.eclipse.microprofile/opentracing.3.0.bnd:
```diff
@@ -0,0 +1,25 @@
+#*******************************************************************************
+# Copyright (c) 2020 IBM Corporation and others.
```
suggest checking the copyright dates throughout

tevans78 commented on dev/io.openliberty.opentracing.2.0.internal/src/io/openliberty/opentracing/internal/OpentracingClientFilter.java:
```diff
@@ -59,8 +61,14 @@
+    }
+    
+    public OpentracingClientFilter(OpentracingFilterHelper helper) {
+    	System.out.println("Is helper null: " + (helper == null));
```
remove System.out.printlns ... replace with proper debug?

tonyreigns replied:
This was just some quick debug for when we were dealing with the providers issue. That's being removed now.

tevans78 reviewed (dismissed):
All new features should be named `io.openliberty...` and initially marked as `noship`

tonyreigns committed:
Feature rename

tonyreigns commented:
#build

Azquelt committed:
mpOpenTracing: update package transform versions

Add transformer rules for bundles which import MP Open Tracing and MP
Rest Client API packages.

Azquelt committed:
mpOpenTracing: fix bundle name reference

Azquelt committed:
opentracing: support for rest client 3.0

With Rest Client 3.0, the OpentracingClientFilter is automatically
applied to rest client calls so we need a different approach to handle
when tracing has been explicitly disabled.

Azquelt committed:
Get OT working with MPRC async

A couple of changes were needed here:
 - use the ManagedExecutorService if available and the user hasn't
   manually specified an executor
 - In the OpentracingClientAsyncWrapper, use a ThreadLocal to pass the
   tracer through to OpentracingClientFilter. This ensures that
   OpentracingClientFilter can get the current Tracer even when a
   managed executor is not used. (There is a TCK test which explicitly
   uses a non-managed executor.)

Azquelt committed:
Add mpOpentracing and mpRestClient to MP 5.0

Add them to the main microprofile feature.

Azquelt committed:
Enable full mpOpenTracing 3.0 TCK

As it now passes.

Azquelt commented:
I've rebased on #18841 as the most recent commits make changes to files introduced in that PR.

Azquelt commented:
Rebased on integration now that #18841 has been merged

Azquelt commented on dev/io.openliberty.microprofile.opentracing.2.0.internal.rest.client/bnd.bnd:
```diff
@@ -38,11 +38,14 @@ Service-Component: \
     com.ibm.ws.jaxrs.2.0.common;version=latest, \
     com.ibm.ws.logging;version=latest, \
     io.openliberty.org.eclipse.microprofile.opentracing.2.0;version=latest, \
+    io.openliberty.org.eclipse.microprofile.opentracing.3.0;version=latest, \
```
Questionable change. Should probably only be opentracing.2.0 on the build path and then the jakarta transformation will make it compatible with opentracing.3.0.

Azquelt commented on dev/io.openliberty.microprofile.opentracing.2.0.internal.rest.client/src/io/openliberty/microprofile/opentracing/internal/rest/client/listener/OpentracingRestClientFilter.java:
```diff
@@ -35,13 +35,17 @@
+        	clientFilter = new OpentracingClientFilter(jaxRsProvider.getOpentracingFilterHelper());
+        	System.out.println("In Rest Client Filter");
+        	
+            //clientFilter = jaxRsProvider.getClientFilter();
```
Needs clean up

Azquelt commented on dev/io.openliberty.microprofile.opentracing.2.0.internal/bnd.bnd:
```diff
@@ -49,6 +53,7 @@ Include-Resource: \
     com.ibm.wsspi.org.osgi.service.component.annotations;version=latest, \
     com.ibm.websphere.org.osgi.service.cm;version=latest, \
     io.openliberty.org.eclipse.microprofile.opentracing.2.0;version=latest, \
+    io.openliberty.org.eclipse.microprofile.opentracing.3.0;version=latest, \
```
Probably only want opentracing.2.0 here.

Azquelt commented on dev/io.openliberty.microprofile.opentracing.3.0.internal.rest.client/bnd.bnd:
```diff
@@ -0,0 +1,46 @@
+-sourcepath: src, resources
+
+-dsannotations:
+
```
Should remove this if there are no components.

Azquelt commented on dev/io.openliberty.microprofile.opentracing.3.0.internal.rest.client/src/io/openliberty/microprofile/opentracing30/internal/rest/client/OpentracingRestClient30Filter.java:
```diff
@@ -0,0 +1,52 @@
+
+    @Override
+    public void filter(ClientRequestContext clientRequestContext) throws IOException {
+        Object invokedMethod = clientRequestContext.getProperty("org.eclipse.microprofile.rest.client.invokedMethod");
```
Is there a constant for this value?

Azquelt replied:
No. Should probably convert all uses of this string to a constant at some point.

Azquelt commented on dev/io.openliberty.microprofile.opentracing.3.0.internal.rest.client/src/io/openliberty/microprofile/opentracing30/internal/rest/client/OpentracingRestClient30Filter.java:
```diff
@@ -0,0 +1,52 @@
+                }
+                
+                // Property read by OpentracingClientFilter
+                clientRequestContext.setProperty(OpentracingClientFilter.CLIENT_FILTER_ENABLED_ID, false);
```
Comment doesn't seem necessary if all references to the property use the constant.

Azquelt commented on dev/io.openliberty.microprofile.opentracing.3.0.internal_fat_tck/fat/src/io/openliberty/microprofile/opentracing/internal/tck/FATSuite.java:
```diff
@@ -0,0 +1,63 @@
+            }
+        }
+        return action;
+    }
```
Can all of this be removed?

Azquelt commented on dev/io.openliberty.microprofile.opentracing.3.0.internal_fat_tck/fat/src/io/openliberty/microprofile/opentracing/internal/tck/OpentracingRestClientTCKLauncher.java:
```diff
@@ -0,0 +1,60 @@
+//    @ClassRule
+//    public static RepeatTests r = RepeatTests.withoutModification()
+//        .andWith(FATSuite.MP_REST_CLIENT("1.3", SERVER_NAME))
+//        .andWith(FATSuite.MP_REST_CLIENT("1.4", SERVER_NAME));
```
Clean up

Azquelt commented on dev/io.openliberty.microprofile.opentracing.3.0.internal_fat_tck/fat/src/io/openliberty/microprofile/opentracing/internal/tck/OpentracingTCKLauncher.java:
```diff
@@ -0,0 +1,52 @@
+    @AllowedFFDC // The tested deployment exceptions cause FFDC so we have to allow for this.
+    public void launchOpentracingTck() throws Exception {
+        // Use default tck-suite.xml
+        MvnUtils.runTCKMvnCmd(server, "io.openliberty.opentracing.2.0.internal_fat", this.getClass() + ":launchOpentracingTck");
```
2.0 -> 3.0

Azquelt commented on dev/io.openliberty.microprofile.opentracing.3.0.internal_fat_tck/fat/src/io/openliberty/microprofile/opentracing/internal/tck/OpentracingTCKLauncherMicroProfile.java:
```diff
@@ -0,0 +1,56 @@
+    @AllowedFFDC // The tested deployment exceptions cause FFDC so we have to allow for this.
+    public void launchOpentracingTckMP() throws Exception {
+        // Use default tck-suite.xml
+        
```
Comment is incorrect

Azquelt commented on dev/io.openliberty.microprofile.opentracing.3.0.internal_fat_tck/opentracing.mock.3.0.bnd:
```diff
@@ -0,0 +1,45 @@
+#*******************************************************************************
+# Copyright (c) 2020 IBM Corporation and others.
```
Copyright date, I guess  `2020, 2021`?

Azquelt commented on dev/io.openliberty.microprofile.opentracing.3.0.internal_fat_tck/publish/servers/OpentracingRestClientTCKServer/server.xml:
```diff
@@ -0,0 +1,26 @@
+<!--Copyright (c) 2020 IBM Corporation and others. All rights reserved. This 
+    program and the accompanying materials are made available under the terms 
```
Copyright date

Azquelt commented on dev/io.openliberty.microprofile.opentracing.3.0.internal_fat_tck/publish/servers/OpentracingTCKServer/server.xml:
```diff
@@ -0,0 +1,25 @@
+<!--Copyright (c) 2020 IBM Corporation and others. All rights reserved. This 
+    program and the accompanying materials are made available under the terms 
```
Copyright date

Azquelt commented on dev/io.openliberty.microprofile.opentracing.3.0.internal_fat_tck/publish/servers/OpentracingTCKServer/server.xml:
```diff
@@ -0,0 +1,25 @@
+
+    <!--Java2 security -->
+    <javaPermission
+        className="java.security.AllPermission" name="*" actions="*" />
```
Add a comment to say why this is needed.

Azquelt replied:
It was needed because there were j2s bugs in MP Rest Client, so I've fixed those and removed this.

Azquelt commented on dev/io.openliberty.microprofile.opentracing.3.0.internal_fat_tck/publish/servers/OpentracingRestClientTCKServer/server.xml:
```diff
@@ -0,0 +1,26 @@
+
+    <!--Java2 security -->
+    <javaPermission className="java.security.AllPermission"
+        name="*" actions="*" />
```
Add a comment to say why this is needed.

Azquelt replied:
It was needed because there were j2s bugs in MP Rest Client, so I've fixed those and removed this.

Azquelt commented on dev/io.openliberty.microprofile.opentracing.3.0.internal_fat_tck/publish/servers/OpentracingTCKServerMicroProfile/server.xml:
```diff
@@ -0,0 +1,27 @@
+<!--Copyright (c) 2020 IBM Corporation and others. All rights reserved. This 
+    program and the accompanying materials are made available under the terms 
```
Copyright date

Azquelt commented on dev/io.openliberty.microprofile.opentracing.3.0.internal_fat_tck/publish/servers/OpentracingTCKServerMicroProfile/server.xml:
```diff
@@ -0,0 +1,27 @@
+
+    <!--Java2 security -->
+    <javaPermission
+        className="java.security.AllPermission" name="*" actions="*" />
```
Add a comment to say why this is needed.

Azquelt replied:
It was needed because there were j2s bugs in MP Rest Client, so I've fixed those and removed this.

Azquelt commented on dev/io.openliberty.microprofile.opentracing.3.0.internal_fat_tck/publish/tckRunner/pom.xml:
```diff
@@ -0,0 +1,71 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<!--
+    Copyright (c) 2020 IBM Corporation and others. All rights reserved.
+    This program and the accompanying materials are made available under the 
```
Copyright date

Azquelt commented on dev/io.openliberty.microprofile.opentracing.3.0.internal_fat_tck/publish/tckRunner/tck/tck-and-rest-client-tck.xml:
```diff
@@ -0,0 +1,23 @@
+<!--Copyright (c) 2020 IBM Corporation and others. All rights reserved.
```
Copyright date

Azquelt commented on dev/io.openliberty.microprofile.opentracing.3.0.internal_fat_tck/publish/tckRunner/tck/tck-suite.xml:
```diff
@@ -0,0 +1,18 @@
+<!--Copyright (c) 2020 IBM Corporation and others. All rights reserved.
```
Copyright date

Azquelt commented on dev/io.openliberty.opentracing.2.0.internal.jaxrs2x/src/io/openliberty/opentracing/internal/jaxrs/OpentracingJaxRsProviderRegister.java:
```diff
@@ -9,82 +9,59 @@
-        instance.compareAndSet(this, null);
         clearContainerFilter();
         clearClientFilter();
     }
```
Each of these methods are one line and they're now only called from activate and deactivate, so it's probably clearer to inline them.

Azquelt commented:
#build - although some parts of the code still need tidying up and reviewing, I want to confirm that all tests pass at this point.

Azquelt commented on dev/io.openliberty.opentracing.2.0.internal/bnd.bnd:
```diff
@@ -32,15 +32,31 @@ Import-Package: \
+    properties:="resources=${app-resources}"
+    
+jakartaeeMe: true
+
```
Somewhere we probably should have a comment to point out that `javax.ws.rs.ext.Providers` is service loaded by `restfulWS-3.0` but a different mechanism is used to add these provides in `jaxrs-2.x`.

andymc12 replied:
Along those lines, it should be `jakarta.ws.rs.ext.Providers` - this bundle uses `javax` and transforms it, but future versions (that aren't transformed) should be `jakarta`.

Azquelt commented on dev/io.openliberty.opentracing.2.0.internal/bnd.bnd:
```diff
@@ -56,7 +72,9 @@ Private-Package: io.openliberty.opentracing.internal.resources
     io.openliberty.org.eclipse.microprofile.config.2.0;version=latest,\
     com.ibm.websphere.javaee.servlet.3.1; version=latest, \
     io.openliberty.org.eclipse.microprofile.opentracing.2.0;version=latest, \
+    io.openliberty.org.eclipse.microprofile.opentracing.3.0;version=latest, \
```
Probably only want opentracing.2.0 here.

Azquelt commented on dev/io.openliberty.opentracing.2.0.internal/src/io/openliberty/opentracing/internal/OpentracingClientFilter.java:
```diff
@@ -43,8 +44,11 @@
 public class OpentracingClientFilter implements ClientRequestFilter, ClientResponseFilter {
     private static final TraceComponent tc = Tr.register(OpentracingClientFilter.class);
+    
+    private static ThreadLocal<Tracer> currentTracer = new ThreadLocal<>();
```
This field and its uses should be reviewed carefully.

Azquelt commented on dev/io.openliberty.opentracing.2.0.internal/src/io/openliberty/opentracing/internal/OpentracingClientFilter.java:
```diff
@@ -84,8 +94,18 @@ void setFilterHelper(OpentracingFilterHelper helper) {
     public void filter(ClientRequestContext clientRequestContext) throws IOException {
         String methodName = "filter(outgoing)";
+        
+        if (!isEnabled(clientRequestContext)) {
+            if (TraceComponent.isAnyTracingEnabled() && tc.isDebugEnabled()) {
+                Tr.debug(this, tc, methodName + " trace disabled for method");
+            }
+            return;
+        }
```
I'm not a huge fan of having this "isEnabled" check here and then a "process" check further down.

It does match the logic from `mpRestClient-2.0` (where the "isEnabled" check is done by a subclass which overrides the filter method), but it would be good to separate out what these two checks are for and why they're different now that they're in the same method.

Azquelt commented on dev/io.openliberty.opentracing.2.0.internal/src/io/openliberty/opentracing/internal/OpentracingClientFilter.java:
```diff
@@ -84,8 +94,18 @@ void setFilterHelper(OpentracingFilterHelper helper) {
+        }
 
-        Tracer tracer = OpentracingTracerManager.getTracer();
+        Tracer tracer = currentTracer.get();
+        if (tracer == null) {
+            tracer = OpentracingTracerManager.getTracer();
+        }
```
Explanatory comment (or refactor for clarity) would be good here.

In restfulWS-3.0, we expect `currentTracer` to be set and we might be running on a non-managed thread. In `jaxrs-2.x` we expect `currentTracer` not to be set but to run on the calling thread, which should be managed so `OpentracingTracerManager.getTracer()` should always work.

Azquelt commented on dev/io.openliberty.opentracing.2.0.internal/src/io/openliberty/opentracing/internal/OpentracingContainerFilter.java:
```diff
@@ -60,9 +62,14 @@
+    
+//    OpentracingContainerFilter(OpentracingFilterHelper helper) {
+//        setFilterHelper(helper);
+//    }
```
Clean up

Azquelt commented on dev/io.openliberty.opentracing.2.0.internal/src/io/openliberty/opentracing/internal/OpentracingFilterHelperProvider.java:
```diff
@@ -0,0 +1,62 @@
+
+    // DSR activation API ...
+
+    protected synchronized void activate(ComponentContext context) {
+        instance.set(this);
+    }
+
+    protected void deactivate(ComponentContext context) {
+        instance.compareAndSet(this, null);
+    }
+
+    public static OpentracingFilterHelperProvider getInstance() {
+        return instance.get();
+    }
+
+    private OpentracingFilterHelper helper;
+
+    @Reference(cardinality = ReferenceCardinality.OPTIONAL, policy = ReferencePolicy.DYNAMIC, policyOption = ReferencePolicyOption.GREEDY)
+    protected synchronized void setOpentracingFilterHelper(OpentracingFilterHelper helper) {
+        this.helper = helper;
+    }
+
+    protected void unsetOpentracingFilterHelper(OpentracingFilterHelper helper) {
+        this.helper = null;
+    }
+
+    public OpentracingFilterHelper getOpentracingFilterHelper() {
+        return helper;
+    }
```
It's unclear to me what `synchronized` is supposed to accomplish here.

Are `unsetOpentracingFilterHelper` and `getOpentracingFilterHelper` supposed to be `synchronized` as well?

It looks like a `volatile` field reference would be sufficient here instead.

fmhwong replied:
`synchronized` isn't needed as `setOpentracingFilterHelper` and `unsetOpentracingFilterHelper` will be called once.  Adding `volatile` may help other threads to get the latest value immediately.

Azquelt commented on dev/io.openliberty.opentracing.3.0.internal.restfulws/src/io/openliberty/opentracing30/internal/restfulws/OpentracingClientAsyncWrapper.java:
```diff
@@ -0,0 +1,79 @@
+
+/**
+ * Wraps the restfulWS client task to capture and apply the current span
+ */
```
Also captures the current tracer.

Azquelt replied:
Done

Azquelt commented on dev/io.openliberty.opentracing.3.0.internal.restfulws/src/io/openliberty/opentracing30/internal/restfulws/OpentracingClientAsyncWrapper.java:
```diff
@@ -0,0 +1,79 @@
+                }
+            }
+        };
+    }
```
This won't capture the current tracer if there isn't also an active span which doesn't seem correct.

Azquelt replied:
Done

Azquelt commented on dev/io.openliberty.org.jboss.resteasy.common/src/io/openliberty/restfulWS/client/internal/ClientAsyncTaskWrapperComponent.java:
```diff
@@ -0,0 +1,86 @@
+                        throw new NullPointerException("ClientAsyncTaskWrapper " + wrapper + " returned null");
+                    }
+                    r = wrapped;
+                } catch (Exception e) {
+                    // FFDC
+                }
```
This is an internal error. Should we have a message here beyond just an FFDC.

Azquelt commented on dev/io.openliberty.org.jboss.resteasy.mprestclient/src/io/openliberty/microprofile/rest/client30/internal/OsgiServices.java:
```diff
@@ -0,0 +1,57 @@
+ *
+ */
+@Component(configurationPolicy = ConfigurationPolicy.IGNORE)
+public class OsgiServices {
```
Needs a comment.

Azquelt replied:
Done

Azquelt commented on dev/io.openliberty.org.jboss.resteasy.mprestclient/src/io/openliberty/microprofile/rest/client30/internal/OsgiServices.java:
```diff
@@ -0,0 +1,57 @@
+
+    public static Optional<ExecutorService> getManagedExecutorService() {
+        return instance.flatMap(tracker -> Optional.ofNullable(tracker.executorService));
+    }
```
Needs javadoc

Azquelt replied:
Would also be better if we didn't create a new `Optional` on every call since we expect this to be called regularly but the `executorService` to change rarely if ever.

Azquelt replied:
Done

Azquelt commented on dev/io.openliberty.org.jboss.resteasy.mprestclient/src/io/openliberty/microprofile/rest/client30/internal/OsgiServices.java:
```diff
@@ -0,0 +1,57 @@
+               cardinality = ReferenceCardinality.OPTIONAL,
+               policy = ReferencePolicy.DYNAMIC,
+               policyOption = ReferencePolicyOption.GREEDY)
+    private volatile ManagedExecutorService executorService;
```
`service` parameter isn't necessary.

Verify that this bundle can rely on the `ManagedExecutorService` interface always being available. We _should_ be able to only depend on `ExecutorService` in here if we need to.

Azquelt replied:
`ManagedExecutorService` is always available when this runs as a feature within liberty: this bundle is included in `io.openliberty.mpRestClient-3.0` which depends on `io.openliberty.restfulWSClient-3.0` which depends on `io.openliberty.concurrent-2.0`

Azquelt commented on dev/io.openliberty.opentracing.2.0.spi/bnd.bnd:
```diff
@@ -23,5 +23,7 @@ Export-Package: io.openliberty.opentracing.spi.tracer;version=2.0
 publish.wlp.jar.suffix: dev/spi/ibm
 
+jakartaeeMe: true
+
```
Double check that adding `jakartaeeMe` to the SPI bundle here is correct...

tonyreigns replied:
Don't believe this is needed, was an early change. Will verify though.

Azquelt reviewed (commented):
Noted a few things that need reviewed or cleaned up.

Azquelt committed:
Review changes to MPRC async support

Azquelt committed:
restclient-3.0 java2security fixes

Azquelt committed:
mpRestClient: Fixup OsgiServices

Last commit prevents it from compiling (through Eclipse is somehow
happy)

Azquelt committed:
mpOpenTracing: TCK review comments

Azquelt committed:
mpOpenTracing: remove 3.0 API from bnd files

Where a bundle is used for openTracing 2.x and 3.x, it should be
compiled against the 2.x API and then transformed to also work with 3.x.

Therefore, only the 2.x API should be on the build path.

Azquelt committed:
opentracing 2.0 review comments

Azquelt committed:
opentracing 3.0 + rest client review comments

Azquelt committed:
opentracing 3.0 + restfulWS review comments

Azquelt commented on dev/io.openliberty.org.jboss.resteasy.mprestclient/src/org/jboss/resteasy/microprofile/client/RestClientBuilderImpl.java:
```diff
@@ -233,7 +242,14 @@ public RestClientBuilder executorService(ExecutorService executor) {
+            // RestClientListeners.get() gets the TCCL
+            listeners = AccessController.doPrivileged((PrivilegedAction<Collection<RestClientListener>>) () -> RestClientListeners.get());
+        }
+        listeners.forEach(listener -> listener.onNewClient(aClass, this));
```
Need to comment our changes in this file since it's an overlay from resteasy

andymc12 replied:
Can we create a Liberty-specific subclass of this with our changes like we did with `LibertyResteasyClientBuilderImpl` ?

Azquelt committed:
mpRestClient: remove static dep on EE concurrent

Ensure that mpRestClient will use an EE managed executor service if one
is available as an OSGi service but doesn't have a hard dependency on
either OSGi or EE concurrency, in case we want to later have it run
standalone.

Azquelt commented:
It's probably a good idea to check through the changes and make sure all new packages have the right `traceGroup` set and that new package names match the trace strings you use for service.

tonyreigns commented:
#build

andymc12 commented on dev/io.openliberty.microprofile.opentracing.3.0.internal_fat_tck/bnd.bnd:
```diff
@@ -0,0 +1,35 @@
+#*******************************************************************************
+# Copyright (c) 2020 IBM Corporation and others.
```
```suggestion
# Copyright (c) 2021 IBM Corporation and others.
```

andymc12 commented on dev/io.openliberty.microprofile.opentracing.3.0.internal_fat_tck/fat/src/io/openliberty/microprofile/opentracing/internal/tck/FATSuite.java:
```diff
@@ -0,0 +1,63 @@
+/*******************************************************************************
+ * Copyright (c) 2020 IBM Corporation and others.
```
```suggestion
 * Copyright (c) 2021 IBM Corporation and others.
```

andymc12 commented on dev/io.openliberty.microprofile.opentracing.3.0.internal_fat_tck/opentracing_fat_tck.bnd:
```diff
@@ -0,0 +1,15 @@
+#*******************************************************************************
+# Copyright (c) 2020 IBM Corporation and others.
```
```suggestion
# Copyright (c) 2021 IBM Corporation and others.
```

andymc12 commented on dev/io.openliberty.microprofile.opentracing.3.0.internal_fat_tck/test-bundles/opentracing.mock/src/io/openliberty/opentracing/internal/mock/OpentracingMockTracer.java:
```diff
@@ -0,0 +1,170 @@
+/*******************************************************************************
+ * Copyright (c) 2020 IBM Corporation and others.
```
```suggestion
 * Copyright (c) 2021 IBM Corporation and others.
```

andymc12 commented on dev/io.openliberty.microprofile.opentracing.3.0.internal_fat_tck/test-bundles/opentracing.mock/src/io/openliberty/opentracing/internal/mock/OpentracingMockTracerFactory.java:
```diff
@@ -0,0 +1,79 @@
+/*******************************************************************************
+ * Copyright (c) 2020 IBM Corporation and others.
```
```suggestion
 * Copyright (c) 2021 IBM Corporation and others.
```

andymc12 commented on dev/io.openliberty.microprofile.opentracing.3.0.internal_fat_tck/test-bundles/opentracing.mock/src/io/openliberty/opentracing/internal/mock/package-info.java:
```diff
@@ -0,0 +1,15 @@
+/*******************************************************************************
+ * Copyright (c) 2020 IBM Corporation and others.
```
```suggestion
 * Copyright (c) 2021 IBM Corporation and others.
```

andymc12 commented on dev/io.openliberty.microprofile.opentracing.3.0.internal_fat_tck/test-bundles/opentracing.mock/src/io/openliberty/opentracing/internal/tracer/OpentracingTracerFactory.java:
```diff
@@ -0,0 +1,27 @@
+/*******************************************************************************
+ * Copyright (c) 2017 IBM Corporation and others.
```
```suggestion
 * Copyright (c) 2021 IBM Corporation and others.
```

andymc12 commented on dev/io.openliberty.microprofile.opentracing.3.0.internal_fat_tck/test-bundles/opentracing.mock/src/io/openliberty/opentracing/internal/tracer/package-info.java:
```diff
@@ -0,0 +1,15 @@
+/*******************************************************************************
+ * Copyright (c) 2020 IBM Corporation and others.
```
```suggestion
 * Copyright (c) 2021 IBM Corporation and others.
```

tonyreigns committed:
Delaying OpenTracingFilterHelper lookup until filte method

Azquelt commented on dev/io.openliberty.org.jboss.resteasy.mprestclient/src/org/jboss/resteasy/microprofile/client/RestClientBuilderImpl.java:
```diff
@@ -310,10 +324,10 @@ public RestClientBuilder executorService(ExecutorService executor) {
-            this.executorService = Executors.newCachedThreadPool();
-            resteasyClientBuilder.executorService(executorService, true);
+            this.executorService = OsgiServices.getManagedExecutorService().orElseGet(Executors::newCachedThreadPool);
+            resteasyClientBuilder.executorService(new AsyncClientExecutorService(executorService), true);
```
I think the error in the latest build is caused by these two lines. Line 330 should pass `false` if using the managed executor service and `true` if using a new cached thread pool.
```suggestion
            Optional<ExecutorService> managedExecutor = OsgiServices.getManagedExecutorService();
            this.executorService = managedExecutor.orElseGet(Executors::newCachedThreadPool);
            resteasyClientBuilder.executorService(new AsyncClientExecutorService(executorService), managedExecutor.isEmpty());
```

tonyreigns committed:
Fixing RestClient errors

tonyreigns committed:
Cleanup and Copyright update

tonyreigns committed:
Initial LibertyRestClientBuilderImpl subclass

tonyreigns commented:
#build

Added the Liberty-specific subclass for RestClientBuilderImpl. Running another build to verify there aren't any issues.

tonyreigns committed:
Fixing conflicts

tonyreigns committed:
Merge branch 'integration' into mpOTJakarta

tonyreigns committed:
Merge branch 'integration' into mpOTJakarta

tonyreigns committed:
Merge branch 'mpOTJakarta' of github.com:tonyreigns/open-liberty into mpOTJakarta

tonyreigns committed:
Updating new transform file post prohibited terms update

fmhwong commented on dev/io.openliberty.microprofile.opentracing.3.0.internal.rest.client/bnd.bnd:
```diff
@@ -0,0 +1,44 @@
+    io.openliberty.org.eclipse.microprofile.rest.client.3.0;version=latest, \
+    io.openliberty.jakarta.restfulWS.3.0;version=latest, \
+    io.openliberty.jakarta.annotation.2.0;version=latest, \
+    io.openliberty.opentracing.2.0.internal;version=latest
```
Should this be `io.openliberty.opentracing.2.0.internal.jakarta`?

fmhwong reviewed (approved):
LGTM

tonyreigns committed:
Cleanup and comments

tonyreigns committed:
Copyright updates and cleanup

tonyreigns commented:
Build after most recent changes: https://wasrtc.hursley.ibm.com:9443/jazz/web/projects/WS-CD#action=com.ibm.team.build.viewResult&id=_TJW6kDw_EeygsrM6ShizkA&tab=com.ibm.team.build.web.ui.internal.editors.result.test.TestProvider

#run-libby-bot
